File: apps/xmtp_sdk_bindgen/runtime/kotlin/SDKCodecs.kt
package uniffi.xmtp_sdk

private fun codecValueError() =
    XmtpException.InvalidArgument(
        ErrorDetails("InvalidArgument", ErrorCategory.INPUT, false, "wrong standard codec value"),
    )

private fun <T : Any> decodePure(
    encoded: EncodedContent,
    take: (StandardContent) -> T?,
): T = take(decodeStandard(encoded)) ?: throw codecValueError()

class TextCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.TEXT)

    override fun encode(value: Any) = encodeStandard(StandardContent.Text(value as String))

    override fun decode(encoded: EncodedContent): Any = decodePure(encoded) { (it as? StandardContent.Text)?.v1 }
}

class MarkdownCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.MARKDOWN)

    override fun encode(value: Any) = encodeStandard(StandardContent.Markdown(value as String))

    override fun decode(encoded: EncodedContent): Any = decodePure(encoded) { (it as? StandardContent.Markdown)?.v1 }
}

class ReadReceiptCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.READ_RECEIPT)

    override fun encode(value: Any): EncodedContent {
        require(value == Unit)
        return encodeStandard(StandardContent.ReadReceipt)
    }

    override fun decode(encoded: EncodedContent): Any =
        decodePure(encoded) {
            if (it is StandardContent.ReadReceipt) Unit else null
        }
}

class ReactionV2Codec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.REACTION)

    override fun encode(value: Any) = encodeStandard(value as? StandardContent.Reaction ?: throw codecValueError())

    override fun decode(encoded: EncodedContent): Any = decodePure(encoded) { it as? StandardContent.Reaction }
}

class AttachmentCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.ATTACHMENT)

    override fun encode(value: Any) = encodeStandard(StandardContent.Attachment(value as Attachment))

    override fun decode(encoded: EncodedContent): Any = decodePure(encoded) { (it as? StandardContent.Attachment)?.v1 }
}

class RemoteAttachmentCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.REMOTE_ATTACHMENT)

    override fun encode(value: Any) = encodeStandard(StandardContent.RemoteAttachment(value as RemoteAttachment))

    override fun decode(encoded: EncodedContent): Any =
        decodePure(encoded) {
            (it as? StandardContent.RemoteAttachment)?.v1
        }
}

class MultiRemoteAttachmentCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.MULTI_REMOTE_ATTACHMENT)

    override fun encode(value: Any) =
        encodeStandard(StandardContent.MultiRemoteAttachment(value as MultiRemoteAttachment))

    override fun decode(encoded: EncodedContent): Any =
        decodePure(encoded) {
            (it as? StandardContent.MultiRemoteAttachment)?.v1
        }
}

class TransactionReferenceCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.TRANSACTION_REFERENCE)

    override fun encode(value: Any) =
        encodeStandard(StandardContent.TransactionReference(value as TransactionReference))

    override fun decode(encoded: EncodedContent): Any =
        decodePure(encoded) {
            (it as? StandardContent.TransactionReference)?.v1
        }
}

class WalletSendCallsCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.WALLET_SEND_CALLS)

    override fun encode(value: Any) = encodeStandard(StandardContent.WalletSendCalls(value as WalletSendCalls))

    override fun decode(encoded: EncodedContent): Any =
        decodePure(encoded) { (it as? StandardContent.WalletSendCalls)?.v1 }
}

class ActionsCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.ACTIONS)

    override fun encode(value: Any) = encodeStandard(StandardContent.Actions(value as Actions))

    override fun decode(encoded: EncodedContent): Any = decodePure(encoded) { (it as? StandardContent.Actions)?.v1 }
}

class IntentCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.INTENT)

    override fun encode(value: Any) = encodeStandard(StandardContent.Intent(value as Intent))

    override fun decode(encoded: EncodedContent): Any = decodePure(encoded) { (it as? StandardContent.Intent)?.v1 }
}

class ReplyCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.REPLY)

    override fun encode(value: Any) = encodeStandard(value as? StandardContent.Reply ?: throw codecValueError())

    override fun decode(encoded: EncodedContent): Any = decodePure(encoded) { it as? StandardContent.Reply }
}

class GroupUpdatedCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.GROUP_UPDATED)

    override fun encode(value: Any) = encodeStandard(StandardContent.GroupUpdated(value as GroupUpdated))

    override fun decode(encoded: EncodedContent): Any =
        decodePure(encoded) { (it as? StandardContent.GroupUpdated)?.v1 }
}

class DeleteMessageCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.DELETE_MESSAGE)

    override fun encode(value: Any) = encodeStandard(value as? StandardContent.DeleteMessage ?: throw codecValueError())

    override fun decode(encoded: EncodedContent): Any = decodePure(encoded) { it as? StandardContent.DeleteMessage }
}

class LeaveRequestCodec : SDKContentCodec {
    override val type get() = standardContentType(StandardContentKind.LEAVE_REQUEST)

    override fun encode(value: Any) = encodeStandard(StandardContent.LeaveRequest(value as LeaveRequest))

    override fun decode(encoded: EncodedContent): Any =
        decodePure(encoded) { (it as? StandardContent.LeaveRequest)?.v1 }
}

File: apps/xmtp_sdk_bindgen/runtime/swift/SDKCodecs.swift
import Foundation

private func codecValueError() -> XmtpError {
    .InvalidArgument(ErrorDetails(code: "InvalidArgument", category: .input, retryable: false, message: "wrong standard codec value"))
}

private func encodePure<T>(_ value: Any, as _: T.Type, wrap: (T) -> StandardContent) throws -> EncodedContent {
    guard let value = value as? T else { throw codecValueError() }
    return try encodeStandard(value: wrap(value))
}

private func decodePure<T>(_ encoded: EncodedContent, take: (StandardContent) -> T?) throws -> Any {
    guard let value = try take(decodeStandard(encoded: encoded)) else { throw codecValueError() }
    return value
}

public struct TextCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .text)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        try encodePure(value, as: String.self, wrap: StandardContent.text)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case let .text(value) = $0 {
                value
            } else {
                nil
            }
        }
    }
}

public struct ReadReceiptCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .readReceipt)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        guard value is Void else { throw codecValueError() }
        return try encodeStandard(value: .readReceipt)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case .readReceipt = $0 {
                ()
            } else {
                nil
            }
        }
    }
}

public struct ReactionV2Codec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .reaction)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        guard let content = value as? StandardContent, case .reaction = content else { throw codecValueError() }
        return try encodeStandard(value: content)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case .reaction = $0 {
                $0
            } else {
                nil
            }
        }
    }
}

public struct AttachmentCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .attachment)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        try encodePure(value, as: Attachment.self, wrap: StandardContent.attachment)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case let .attachment(value) = $0 {
                value
            } else {
                nil
            }
        }
    }
}

public struct RemoteAttachmentCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .remoteAttachment)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        try encodePure(value, as: RemoteAttachment.self, wrap: StandardContent.remoteAttachment)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case let .remoteAttachment(value) = $0 {
                value
            } else {
                nil
            }
        }
    }
}

public struct MultiRemoteAttachmentCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .multiRemoteAttachment)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        try encodePure(value, as: MultiRemoteAttachment.self, wrap: StandardContent.multiRemoteAttachment)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case let .multiRemoteAttachment(value) = $0 {
                value
            } else {
                nil
            }
        }
    }
}

public struct TransactionReferenceCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .transactionReference)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        try encodePure(value, as: TransactionReference.self, wrap: StandardContent.transactionReference)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case let .transactionReference(value) = $0 {
                value
            } else {
                nil
            }
        }
    }
}

public struct ReplyCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .reply)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        guard let content = value as? StandardContent, case .reply = content else { throw codecValueError() }
        return try encodeStandard(value: content)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case .reply = $0 {
                $0
            } else {
                nil
            }
        }
    }
}

public struct GroupUpdatedCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .groupUpdated)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        try encodePure(value, as: GroupUpdated.self, wrap: StandardContent.groupUpdated)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case let .groupUpdated(value) = $0 {
                value
            } else {
                nil
            }
        }
    }
}

public struct DeleteMessageCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .deleteMessage)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        guard let content = value as? StandardContent, case .deleteMessage = content else { throw codecValueError() }
        return try encodeStandard(value: content)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case .deleteMessage = $0 {
                $0
            } else {
                nil
            }
        }
    }
}

public struct LeaveRequestCodec: SDKContentCodec {
    public init() {}
    public var type: ContentTypeID {
        standardContentType(kind: .leaveRequest)
    }

    public func encode(_ value: Any) throws -> EncodedContent {
        try encodePure(value, as: LeaveRequest.self, wrap: StandardContent.leaveRequest)
    }

    public func decode(_ encoded: EncodedContent) throws -> Any {
        try decodePure(encoded) {
            if case let .leaveRequest(value) = $0 {
                value
            } else {
                nil
            }
        }
    }
}
